docs/ovs-pipeline.md

```diff
@@ -0,0 +1,561 @@
+destination MAC address and that the packet must be forwarded. In the last table
+of the pipeline ([L2ForwardingOutTable]), we will drop all packets for which
+this bit is not set. We also use the NXM_NX_REG1 register to store the egress
+port for the packet, which will be used as a parameter to the `output` OpenFlow
+action in [L2ForwardingOutTable].
+
+All packets - whether they have a matching dmac entry or not - are then
+resubmitted to the next table, [IngressRuleTable].
+
+What about L2 multicast / broadcast traffic? ARP requests will never reach this
+table, as they will be handled by the OpenFlow `normal` action in the
+[ArpResponderTable]. As for the rest, if it is IP traffic, it will hit the
+"last" flow in this table and be resubmitted to [IngressRuleTable]. Assuming it
+makes it to the last table of the pipeline ([L2ForwardingOutTable]), it will be
+dropped there since bit 16 of the NXM_NX_REG0 will not be set. Traffic which is
+non-ARP and non-IP (assuming any can be received by the switch) is actually
+dropped much earlier in the pipeline ([SpoofGuardTable]). In the future, we may
+need to support more cases for L2 multicast / broadcast traffic.
+
+### IngressRuleTable (90)
+
+This table is very similar to [EgressRuleTable], but implements ingress rules
+for Network Policies. Once again, you will need to keep mind the Network Policy
+[specification](#network-policy-implementation) that we are using. We have 2
+Pods running on the same Node, with IP addresses 10.10.1.2 to 10.10.1.3. They
+are allowed to talk to each other using TCP on port 80, but nothing else.
+
+If you dump the flows for this table, you should see something like this:
+```
+1. table=90, priority=210,ct_state=-new+est,ip actions=resubmit(,110)
+2. table=90, priority=210,ip,nw_src=10.10.1.1 actions=resubmit(,110)
+3. table=90, priority=200,ip,nw_src=10.10.1.2 actions=conjunction(1,1/3)
+4. table=90, priority=200,ip,nw_src=10.10.1.3 actions=conjunction(1,1/3)
+5. table=90, priority=200,ip,reg1=0x3 actions=conjunction(1,2/3)
+6. table=90, priority=200,ip,reg1=0x4 actions=conjunction(1,2/3)
+7. table=90, priority=200,tcp,tp_dst=80 actions=conjunction(1,3/3)
+8. table=90, priority=190,conj_id=1,ip actions=resubmit(,110)
+9. table=90, priority=80,ip actions=resubmit(,100)
+```
+
+As for [EgressRuleTable], flow 1 (highest priority) ensures that for established
+connections - as a remainder all connections are committed in
+[ConntrackStateTable] - packets go straight to [L2ForwardingOutTable], with no
+other match required.
+
+Flow 2 ensures that traffic from the local gateway cannot be dropped because of
+Network Policies. This ensures that K8s [liveness
+probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)
+can go through.
+
+The rest of the flows read as follows: if the source IP address is in set
+{10.10.1.2, 10.10.1.3}, and the destination port is in the set {3, 4} (which
```

Comment:
maybe "the destination OF port" to avoid confusion.

Reply:
+1